src/System.Management.Automation/engine/CommandSearcher.cs

```diff
@@ -635,7 +635,41 @@ private CommandInfo GetInfoFromPath(string path)
                     break;
                 }
 
-                if (String.Equals(extension, StringLiterals.PowerShellScriptFileExtension, StringComparison.OrdinalIgnoreCase))
+                // handle case where we are called by the OS to handle a shebang script using us as the interpreter
+                bool isShebangScript = false;
+                try
+                {
+                    using(FileStream fileStream = new FileStream(path, FileMode.Open, FileAccess.Read))
```

Comment:
There's a lot of code! Can we just read the line (with length < Path.Max) and check Shebang with `RegEx.IsMatch()`? And it seems we're going to have less trouble with BOM.